### PR TITLE
Simple Delegation follow up

### PR DESCRIPTION
**Pull request by katamarinaki**

### CODE REVIEW

File: apps/voting/contracts/Voting.sol [hunk 1]
```diff
@@ -122,79 +130,74 @@ contract Voting is IForwarder, AragonApp {
...
         _removeDelegatedAddressFor(prevDelegate, msgSender);
-        delegates[msgSender] = address(0);
+        delegates[msgSender] = Delegate(address(0), 0);
 
         emit DelegateSet(msgSender, prevDelegate, address(0));
     }
 
     function _addDelegatedAddressFor(address _delegate, address _voter) internal {
-        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
-        require(_voter != address(0), ERROR_ZERO_ADDRESS_PASSED);
```
Comment (kadmil) [inline apps/voting/contracts/Voting.sol:163]: why no zero checks? or are those used with `require`s before that every time?

File: apps/voting/contracts/Voting.sol [hunk 2]
```diff
@@ -122,79 +130,74 @@ contract Voting is IForwarder, AragonApp {
...
-        for (uint256 i = 0; i < length; i++) {
-            if (delegatedVoters[_delegate].addresses[i] == _voter) {
-                delegatedVoters[_delegate].addresses[i] = delegatedVoters[_delegate].addresses[length - 1];
-                delete delegatedVoters[_delegate].addresses[length - 1];
-                delegatedVoters[_delegate].addresses.length--;
-                break;
-            }
+        delegatedVoters[_delegate].addresses[voterIndex] = delegatedVoters[_delegate].addresses[length - 1];
+        delegates[delegatedVoters[_delegate].addresses[length - 1]].voterIndex = voterIndex;
+        delete delegatedVoters[_delegate].addresses[length - 1];
```
Comment (kadmil) [inline apps/voting/contracts/Voting.sol]: needs a separate unittest checking "last delegate changed / last delegate removed" for single- and multiple-delegate lists

File: apps/voting/contracts/Voting.sol [hunk 3]
```diff
@@ -122,79 +130,74 @@ contract Voting is IForwarder, AragonApp {
...
+        delegatedVoters[_delegate].addresses.length--;
+    }
+
+    function getDelegatedVoters(address _delegate, uint96 offset, uint96 count) public view returns (address[] memory) {
```
Comment (kadmil) [inline apps/voting/contracts/Voting.sol]: add default values maybe? (0, 100) or something, depending on gas requirements

File: apps/voting/contracts/Voting.sol [hunk 4]
```diff
@@ -290,30 +293,44 @@ contract Voting is IForwarder, AragonApp {
...
+        uint256 skippedVotersCount;
+
+        for (uint256 i = 0; i < length; ++i) {
+            address voteFor_ = _voteForList[i];
+            uint256 votingPower = token.balanceOfAt(voteFor_, vote_.snapshotBlock);
+            VoterState state = vote_.voters[voteFor_];
+            if (_canVoteFor(msgSender, voteFor_) && votingPower > 0 && state != VoterState.Yea && state != VoterState.Nay) {
+                _vote(_voteId, _supports, voteFor_, true);
+            } else {
+                emit VoteForMultipleSkippedFor(_voteId, msgSender, voteFor_, _supports);
```
Comment (kadmil) [inline apps/voting/contracts/Voting.sol]: better emit single "summary" event after the cycle imo

File: apps/voting/contracts/Voting.sol [hunk 5]
```diff
@@ -45,6 +44,8 @@ contract Voting is IForwarder, AragonApp {
     string private constant ERROR_DELEGATE_SAME_AS_PREV = "VOTING_DELEGATE_SAME_AS_PREV";
     string private constant ERROR_MAX_DELEGATED_VOTERS_REACHED = "VOTING_MAX_DELEGATED_VOTERS_REACHED";
```
Comment (bulbozaur) [inline apps/voting/contracts/Voting.sol:48]: unused constant

File: apps/voting/contracts/Voting.sol (hunk 1 above)
Comment (katamarinaki) [reply apps/voting/contracts/Voting.sol:163]: The `_delegate` is checked in require [before](https://github.com/katamarinaki/aragon-apps/pull/2/files/337b031e11ad56f5a228d1fe94e9012d195d69f4#diff-b6a08f95413f3092266f49a1084493af6d2596fc3118b0cede9af64dd8d3c402R134) the call, the `_voter` is `msg.sender` in the only one call of this fn, so it's not worth checking imo.

File: apps/voting/contracts/Voting.sol (hunk 5 above)
Comment (katamarinaki) [reply apps/voting/contracts/Voting.sol:48]: Thanks, fixed.

File: apps/voting/contracts/Voting.sol [hunk 6]
```diff
@@ -122,79 +129,74 @@ contract Voting is IForwarder, AragonApp {
...
         uint256 length = delegatedVoters[_delegate].addresses.length;
-        for (uint256 i = 0; i < length; i++) {
-            if (delegatedVoters[_delegate].addresses[i] == _voter) {
-                delegatedVoters[_delegate].addresses[i] = delegatedVoters[_delegate].addresses[length - 1];
-                delete delegatedVoters[_delegate].addresses[length - 1];
-                delegatedVoters[_delegate].addresses.length--;
-                break;
-            }
+        delegatedVoters[_delegate].addresses[voterIndex] = delegatedVoters[_delegate].addresses[length - 1];
+        delegates[delegatedVoters[_delegate].addresses[length - 1]].voterIndex = voterIndex;
```
Comment (Psirex) [inline apps/voting/contracts/Voting.sol]: - Cache the `delegatedVoters[_delegate].addresses[length - 1]` in the variable to reduce storage reads.
- Add the if statement which checks that the voterIndex is not equal to the length - 1. If it's the last in the array, the swap is not needed.
- Seems like explicit `delete` of address is redundant. The array truncate must be enough there.

File: apps/voting/contracts/Voting.sol [hunk 7]
```diff
@@ -122,79 +129,74 @@ contract Voting is IForwarder, AragonApp {
...
+        uint256 votingPower = token.balanceOfAt(msgSender, getBlockNumber64() - 1);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
         if (prevDelegate != address(0)) {
             _removeDelegatedAddressFor(prevDelegate, msgSender);
         }
 
         _addDelegatedAddressFor(_delegate, msgSender);
-        delegates[msgSender] = _delegate;
+        delegates[msgSender] = Delegate(_delegate, uint96(delegatedVoters[_delegate].addresses.length - 1));
```
Comment (Psirex) [inline apps/voting/contracts/Voting.sol]: It's safer to check that overflow does not happen there

File: apps/voting/contracts/Voting.sol [hunk 8]
```diff
@@ -122,79 +129,74 @@ contract Voting is IForwarder, AragonApp {
...
-                delegatedVoters[_delegate].addresses.length--;
-                break;
-            }
+        delegatedVoters[_delegate].addresses[voterIndex] = delegatedVoters[_delegate].addresses[length - 1];
+        delegates[delegatedVoters[_delegate].addresses[length - 1]].voterIndex = voterIndex;
+        delete delegatedVoters[_delegate].addresses[length - 1];
+        delegatedVoters[_delegate].addresses.length--;
+    }
+
+    function getDelegatedVoters(address _delegate, uint96 offset, uint96 count) public view returns (address[] memory) {
```
Comment (Psirex) [inline apps/voting/contracts/Voting.sol]: Usually, we pass args as `uint256` and check that there is no overflow in the function body.

File: apps/voting/contracts/Voting.sol [hunk 9]
```diff
@@ -122,79 +129,74 @@ contract Voting is IForwarder, AragonApp {
...
+        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
+        require(count > 0 && offset >= 0 && offset + count <= delegatedVoters[_delegate].addresses.length, ERROR_INVALID_OFFSET_AND_COUNT);
```
Comment (Psirex) [inline apps/voting/contracts/Voting.sol]: `offset + count` may overflow. It's better to use `uint256` there and safe math. Also, offset can't be less than zero, can't it?

File: apps/voting/contracts/Voting.sol [hunk 10]
```diff
@@ -290,30 +292,44 @@ contract Voting is IForwarder, AragonApp {
     */
     function vote(uint256 _voteId, bool _supports, bool _executesIfDecided_deprecated) external voteExists(_voteId) {
         require(_canVote(_voteId, msg.sender), ERROR_CAN_NOT_VOTE);
-        require(!_supports || _getVotePhase(votes[_voteId]) == VotePhase.Main, ERROR_CAN_NOT_VOTE);
+        require(_canCastYeaVote(_voteId, _supports), ERROR_CAN_NOT_VOTE);
```
Comment (Psirex) [inline apps/voting/contracts/Voting.sol]: This function is a bit confusing. It returns true, meaning that the voter can cast a yea vote when he actually votes against it, passing `_supports` equal to false. A bit illogical

File: apps/voting/contracts/Voting.sol (hunk 6 above)
Comment (ujenjt) [reply apps/voting/contracts/Voting.sol]: It's better to keep delete btw, cause it nullifies the slot

File: apps/voting/contracts/Voting.sol (hunk 10 above)
Comment (katamarinaki) [reply apps/voting/contracts/Voting.sol]: Renamed the function, thanks for pointing out!

File: apps/voting/contracts/Voting.sol (hunk 9 above)
Comment (katamarinaki) [reply apps/voting/contracts/Voting.sol]: Rewrote this check a bit, thanks!

**Comment by katamarinaki** [comment]
Closed in favor or https://github.com/lidofinance/aragon-apps/pull/39

**Pull request closed without merging**